client/main.js

```diff
@@ -136,30 +131,95 @@ const processKeyPoints = (combinedKeyPoints) => {
+  }
+
+  // Render FaceMesh
+  const faceMeshes = combinedFeatures[0];
+  let facePoints;
+  if (faceMeshes !== undefined && faceMeshes.length > 0) {
+    facePoints = faceMeshes[0].scaledMesh;
+  }
+
+  // Render HandPose
+  const handMeshes = combinedFeatures[1];
+  let handPoints;
+  let handAnnotations;
+  if (handMeshes) {
+    handPoints = handMeshes[0].landmarks;
+    handAnnotations = handMeshes[0].annotations;
+  }
+  return [facePoints, handPoints, handAnnotations];
+};
+
 /*
  * Each call to update carries out all key point computations,
  * any enabled processing, and draws the next frame. If no errors
  * occur, asks the runtime to be called again on next frame update.
  */
-const startEngine = async (canvas, video) => {
-  (function update() {
-    computeCombinedKeyPoints(video)
-      .then((combinedFeatures) => drawFrame(canvas, video, combinedFeatures))
-      .then((combinedKeyPoints) => processKeyPoints(combinedKeyPoints))
-      .then(() => requestAnimationFrame(update))
-      .catch((err) =>
-        // eslint-disable-next-line no-console
-        console.error('Could not compute and render frame: ', err)
-      );
-  })();
-};
+const computeFrame = async () =>
+  computeCombinedKeyPoints(state.video)
+    .then((combinedFeatures) => extractPoints(combinedFeatures))
+    .then(([facePoints, handPoints, handAnnotations]) => {
+      // Don't need to draw on canvas if tab is NOT active (i.e. visible)
+      if (state.isActiveTab) {
+        drawFrame(
+          state.canvas,
+          state.video,
+          facePoints,
+          handPoints,
+          handAnnotations
+        );
+      }
+      return processKeyPoints(facePoints, handPoints);
+    })
+    .then(() => {
+      // requestAnimationFrame only works on an active tab
+      if (state.isActiveTab) {
+        state.timerId = requestAnimationFrame(computeFrame);
+      }
+      return Promise.resolve();
+    })
+    .catch((err) =>
+      // eslint-disable-next-line no-console
+      console.error('Could not compute and render frame: ', err)
+    );
 
 /*
  * Initializes video feed models and tf.js runtime.
  * Failures should be fatal.
  */
 const initialize = async () => Promise.all([initializeModel(state), initDom()]);
 
+/*
+ * Handles when user switches tabs and initiates the process with appropriate loop.
+ * No need for canvas when on a DIFFERENT tab.
+ * Should only use requestAnimationFrame when tab=active.
+ * Also responsible for the stopping the previous timer.
+ */
+const handleVisibilityChange = () => {
+  if (document.visibilityState === 'hidden') {
+    if (state.timerId) {
+      cancelAnimationFrame(state.timerId);
+    }
+    state.isActiveTab = false;
+    state.canvas.style.display = 'none';
+    state.timerId = setInterval(computeFrame, FRAMES_PER_SECOND);
+  } else {
+    if (state.timerId) {
+      clearTimeout(state.timerId);
+    }
+    state.canvas.style.display = 'inline-block';
+    state.isActiveTab = true;
```

Comment:
Nit: Wouldn't it be better to set `state.isActiveTab` with the `document.visibilityState === 'hidden'`?
This way, the function is lighter (2LOCs less) and the condition would also be clearer.

Reply:
Yep yep sure. I've made a change, please have a look and let me know if this is better!

Also found our that I was using the incorrect function for stopping the interval so I addressed that as well!